File: impl/web/oberonfs.js
// Чтение файловой системы Оберона прямо из образа, лежащего в памяти машины.
// Порт tools/oberonfs.py; раскладка снята с исходников системы:
//   Kernel.Mod:  сектор adr лежит по смещению (adr DIV 29 - 1) * 1024
//   FileDir.Mod: DirPage = mark, m, p0, fill[52], e[24]; запись = имя 32, adr, p
//                FileHeader = mark, name[32], aleng, bleng, date, ext[12], sec[64]
//   Files.Mod:   длина = aleng * 1024 + bleng - 352
//
// Нужен лабораториям, чтобы проверять результат по ДИСКУ, а не по надписи
// на экране: «скомпилировалось» и «файл на диске изменился» — разные вещи.

const SS = 1024, HS = 352, DIRROOT = 29;

/** Разбор объектного файла (.rsc). Раскладка та же, что в tools/rsc.py,
    снята с ORTool.DecObj: имя, ключ, версия, размер, импорты, дескрипторы
    типов, размер данных, строки, КОД. Нужен лабораториям, чтобы сравнивать
    размер порождённого кода и ключ интерфейса. */
export function parseRsc(bytes) {
  let i = 0;
  const str = () => { let s = ''; while (bytes[i]) s += String.fromCharCode(bytes[i++]); i++; return s; };
  const int = () => { const v = bytes[i] | bytes[i+1]<<8 | bytes[i+2]<<16 | bytes[i+3]<<24; i += 4; return v; };
  const name = str(), key = int() >>> 0, version = bytes[i++], size = int();
  const imports = [];
  for (let n = str(); n; n = str()) imports.push([n, int() >>> 0]);
  const td = int() / 4; i += td * 4;
  const datasize = int();
  const slen = int(); i += slen;
  const nwords = int();
  const code = new Uint32Array(nwords);
  for (let k = 0; k < nwords; k++) code[k] = int() >>> 0;
  return { name, key, version, size, imports, datasize, codeWords: nwords, code };
}

/** Текст Оберона из файла.

    Файлы, сохранённые редактором, лежат НЕ в виде голого ASCII: первый байт —
    метка формата (F1H), затем смещение текста, затем описания кусков со
    шрифтами. Наивное чтение даёт мусор в начале, а концы строк — возврат
    каретки, а не перевод. Разбор по Texts.Mod: Open читает метку, Load —
    смещение и куски. */
export function readText(bytes) {
  if (bytes[0] !== 0xF1) return new TextDecoder('latin1').decode(bytes);   // простой ASCII
  const off = bytes[1] | bytes[2] << 8 | bytes[3] << 16 | bytes[4] << 24;
  return new TextDecoder('latin1').decode(bytes.subarray(off)).replace(/\r/g, '\n');
}

export class OberonFS {
  constructor(machine) { this.m = machine; }

  secOff(adr) { return ((adr / DIRROOT | 0) - 1) * SS; }
  word(adr, i) { return this.m.disk(this.secOff(adr) + i * 4); }

  byte(adr, i) {
    const w = this.m.disk(this.secOff(adr) + (i & ~3));
    return (w >>> ((i & 3) * 8)) & 0xFF;
  }

  name(adr, off) {
    let s = '';
    for (let i = 0; i < 32; i++) {
      const c = this.byte(adr, off + i);
      if (!c) break;
      s += String.fromCharCode(c);
    }
    return s;
  }

  *entries(adr = DIRROOT) {
    if (!adr) return;
    const m = this.word(adr, 1) | 0, p0 = this.word(adr, 2) | 0;
    yield* this.entries(p0);
    for (let i = 0; i < m; i++) {
      const base = 64 + i * 40;
      const nm = this.name(adr, base);
      const hdr = this.m.disk(this.secOff(adr) + base + 32) | 0;
      const p = this.m.disk(this.secOff(adr) + base + 36) | 0;
      yield [nm, hdr];
      yield* this.entries(p);
    }
  }

  files() {
    const out = new Map();
    for (const [n, h] of this.entries()) out.set(n, h);
    return out;
  }

  length(hdr) {
    const aleng = this.m.disk(this.secOff(hdr) + 36) | 0;
    const bleng = this.m.disk(this.secOff(hdr) + 40) | 0;
    return aleng * SS + bleng - HS;
  }

  read(hdr) {
    const base = this.secOff(hdr);
    const aleng = this.m.disk(base + 36) | 0, bleng = this.m.disk(base + 40) | 0;
    const len = aleng * SS + bleng - HS;
    const out = new Uint8Array(len);
    let o = 0;
    const copy = (off, n) => {
      for (let i = 0; i < n && o < len; i += 4) {
        const w = this.m.disk(off + i);
        for (let b = 0; b < 4 && o < len; b++) out[o++] = (w >>> (b * 8)) & 0xFF;
      }
    };
    copy(base + HS, SS - HS);
    for (let i = 1; i <= aleng; i++) {
      let a;
      if (i < 64) a = this.m.disk(base + 96 + i * 4) | 0;
      else {
        const ext = this.m.disk(base + 48 + (((i - 64) / 256) | 0) * 4) | 0;
        a = this.m.disk(this.secOff(ext) + ((i - 64) % 256) * 4) | 0;
      }
      copy(this.secOff(a), SS);
    }
    return out;
  }
}

File: impl/tools/oberonfs.py
#!/usr/bin/env python3
"""Чтение файловой системы Оберона прямо из образа диска.

Раскладка снята с исходников системы, а не угадана:

  Kernel.Mod:  GetSector(src) -> src DIV 29, затем *2 + FSoffset (80000H) в
               512-байтных блоках SD. Эмулятор диска вычитает 80002H
               (tb/disk/disk.c), поэтому сектор adr лежит в образе по
               смещению (adr DIV 29 - 1) * 1024.

  FileDir.Mod: DirPage = mark, m, p0, fill[52], e[24] — 64 байта заголовка и
               24 записи по 40 байт (имя 32, adr 4, p 4). Это B-дерево:
               p0 — левый потомок, p каждой записи — правый.
               FileHeader = mark, name[32], aleng, bleng, date, ext[12],
               sec[64] — ровно 352 байта (HeaderSize), дальше данные.

  Files.Mod:   Length(f) = aleng * 1024 + bleng - 352.

Нужен, чтобы доказать неподвижную точку пересборки: сравнить объектные файлы
до и после побайтово, а не по числам на экране.
"""
import struct, sys

SS, HS, DIRROOT = 1024, 352, 29


class Image:
    def __init__(self, path):
        self.d = open(path, "rb").read()

    def sector(self, adr):
        k = adr // DIRROOT
        off = (k - 1) * SS
        return self.d[off:off + SS]

    def entries(self, adr=DIRROOT):
        """Обход B-дерева каталога: имя -> адрес заголовка файла."""
        if adr == 0:
            return
        s = self.sector(adr)
        m, p0 = struct.unpack_from("<ii", s, 4)
        yield from self.entries(p0)
        for i in range(m):
            base = 64 + i * 40
            name = s[base:base + 32].split(b"\0")[0].decode("latin-1")
            hdr, p = struct.unpack_from("<ii", s, base + 32)
            yield name, hdr
            yield from self.entries(p)

    def date(self, hdr):
        """Отметка времени из заголовка файла (FileDir.Mod: поле date).

        Отличает «пересобрано и совпало побайтово» от «файл не трогали»:
        побайтовое сравнение эти два случая не различает, а это ровно та
        разница, из-за которой проверка могла бы тихо пройти на несобранном.
        """
        return struct.unpack_from("<i", self.sector(hdr), 44)[0]

    def read(self, hdr):
        h = self.sector(hdr)
        aleng, bleng = struct.unpack_from("<ii", h, 36)
        sec = struct.unpack_from("<64i", h, 96)
        length = aleng * SS + bleng - HS
        out = bytearray(h[HS:])                       # хвост первого сектора
        for i in range(1, aleng + 1):
            if i < 64:
                a = sec[i]
            else:                                     # длинные файлы: индексные страницы
                ext = struct.unpack_from("<12i", h, 48)
                page = self.sector(ext[(i - 64) // 256])
                a = struct.unpack_from("<i", page, ((i - 64) % 256) * 4)[0]
            out += self.sector(a)
        return bytes(out[:length])

    def files(self):
        return dict(self.entries())


if __name__ == "__main__":
    img = Image(sys.argv[1])
    fs = img.files()
    if len(sys.argv) > 2:
        for name in sys.argv[2:]:
            sys.stdout.buffer.write(img.read(fs[name]))
    else:
        print(f"файлов: {len(fs)}")
        for n in sorted(fs)[:8]:
            print(f"  {n:24s} {len(img.read(fs[n])):8d} байт")
